/**
 * \file cannyParametersControlWidget.cpp
 * \attention This file is automatically generated and should not be in general modified manually
 *
 * \date MMM DD, 20YY
 * \author autoGenerator
 */

#include "cannyParametersControlWidget.h"
#include "ui_cannyParametersControlWidget.h"
#include "qSettingsGetter.h"
#include "qSettingsSetter.h"


CannyParametersControlWidget::CannyParametersControlWidget(QWidget *parent, bool _autoInit, QString _rootPath)
    : FilterParametersControlWidgetBase(parent)
    , mUi(new Ui::CannyParametersControlWidget)
    , autoInit(_autoInit)
    , rootPath(_rootPath)
{
    mUi->setupUi(this);

    QObject::connect(mUi->shouldEdgeDetectCheckBox, SIGNAL(stateChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->minimumThresholdSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
    QObject::connect(mUi->maximumThresholdSpinBox, SIGNAL(valueChanged(int)), this, SIGNAL(paramsChanged()));
}

CannyParametersControlWidget::~CannyParametersControlWidget()
{

    delete mUi;
}

void CannyParametersControlWidget::loadParamWidget(WidgetLoader &loader)
{
    CannyParameters *params = createParameters();
    loader.loadParameters(*params, rootPath);
    setParameters(*params);
    delete params;
}

void CannyParametersControlWidget::saveParamWidget(WidgetSaver  &saver)
{
    CannyParameters *params = createParameters();
    saver.saveParameters(*params, rootPath);
    delete params;
}

 /* Composite fields are NOT supported so far */
void CannyParametersControlWidget::getParameters(CannyParameters& params) const
{

    params.setShouldEdgeDetect (mUi->shouldEdgeDetectCheckBox->isChecked());
    params.setMinimumThreshold (mUi->minimumThresholdSpinBox->value());
    params.setMaximumThreshold (mUi->maximumThresholdSpinBox->value());

}

CannyParameters *CannyParametersControlWidget::createParameters() const
{

    /**
     * We should think of returning parameters by value or saving them in a preallocated place
     **/


    CannyParameters *result = new CannyParameters(
          mUi->shouldEdgeDetectCheckBox->isChecked()
        , mUi->minimumThresholdSpinBox->value()
        , mUi->maximumThresholdSpinBox->value()
    );
    return result;
}

void CannyParametersControlWidget::setParameters(const CannyParameters &input)
{
    // Block signals to send them all at once
    bool wasBlocked = blockSignals(true);
    mUi->shouldEdgeDetectCheckBox->setChecked(input.shouldEdgeDetect());
    mUi->minimumThresholdSpinBox->setValue(input.minimumThreshold());
    mUi->maximumThresholdSpinBox->setValue(input.maximumThreshold());
    blockSignals(wasBlocked);
    emit paramsChanged();
}

void CannyParametersControlWidget::setParametersVirtual(void *input)
{
    // Modify widget parameters from outside
    CannyParameters *inputCasted = static_cast<CannyParameters *>(input);
    setParameters(*inputCasted);
}
